### PR TITLE
uncache v1 state.

### PR DESCRIPTION
**Pull request by EmilLa1**
## Purpose

Addresses https://github.com/Concordium/concordium-node/issues/462

This PR evicts smart contract state from memory by simply not caching the state. 
 
## Changes

- Evicting state references by converting them to 'Disk' as opposed to 'Cached' references.
- Removed `cache_persistent_state_v1` as sc state should no longer be cached on startup.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs [hunk 1]
```diff
@@ -450,36 +450,24 @@ impl<V> CachedRef<V> {
...
-                    reference,
-                };
-            } else {
-                // this is unreachable since we hold a mutable reference to the cached value
-                // and we know the value was a purely in-memory one
-                unsafe { std::hint::unreachable_unchecked() }
-            }
+            };
         }
     }
```
Comment (abizjak) [inline smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs:465]: Should you not also uncache this if it is "cached"? And set it to the Disk variant?

File: smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs [hunk 2]
```diff
@@ -3271,14 +3252,19 @@ impl Node {
...
-                {
-                    return false;
-                }
+                match &value.data {
+                    CachedRef::Disk {
+                        reference: _,
+                    } => return false,
+                    // [Cached] or [Memory]
+                    _ => return true,
+                };
```
Comment (abizjak) [inline smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs:3261]:
```suggestion
                !matches!(&value.data, CachedRef::Disk{..})
```

Comment (EmilLa1) [reply smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs:3261]: 🥇

File: smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs (hunk 1 above)
Comment (EmilLa1) [reply smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs:465]: Yes, that is correct. 👍

File: smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs [hunk 3]
```diff
@@ -496,29 +479,22 @@ impl<V> CachedRef<V> {
...
                 reference.store(buf)
             }
             CachedRef::Cached {
-                reference,
-                ..
-            } => reference.store(buf),
+                reference: _,
+                value,
+            } => {
+                let reference = backing_store.store_raw(value.as_ref())?;
```
Comment (abizjak) [inline smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs:492]: I don't think you want to be storing the value again, right? Why not reuse the reference you already have?

Comment (EmilLa1) [reply smart-contracts/wasm-chain-integration/src/v1/trie/low_level.rs]: Excellent observation. Fixed now.

**Review by abizjak — APPROVED** [review]

**Pull request merged**